**Câu 101:** Viết chương trình in ra danh sách list các mã byte là của một chuỗi.

In [1]:
x = b'Abc'
print(list(x))

[65, 98, 99]


**Câu 102:** Viết hàm tính tỷ lệ số âm , số dương, số zeros có trong mảng được nhập từ bàn phím cách nhau bởi dấu cách.

In [2]:
def plusMinus(arr):
    count_1=0;count_2=0;count_3=0
    for i in arr:
        if i>0:
            count_1+=1
        elif i<0:
            count_2+=1
        else:
            count_3+=1
    print("Phần trăm số dương trong mảng là: {0:.6f}".format(count_1/len(arr)))
    print("Phần trăm số âm trong mảng là: {0:.6f}".format(count_2/len(arr)))
    print("Phần trong số zeros trong mảng là: {0:.6f}".format(count_3/len(arr)))
n=int(input("Nhập số lượng phần tử cả mảng:"))
arr=[]
for i in input().split(" "):
    arr.append(int(i))
plusMinus(arr)

Nhập số lượng phần tử cả mảng:6
-2 2 0 4 2 3
Phần trăm số dương trong mảng là: 0.666667
Phần trăm số âm trong mảng là: 0.166667
Phần trong số zeros trong mảng là: 0.166667


**Câu 103:** Viết hàm tìm số lần để cộng các mẫu(10, 20, 50, 100, 200, 500) thành số n:

<div style="padding: 15px; border: 2px solid transparent; border-color: transparent; margin-bottom: 5px; margin-top: 10px;border-radius: 2px; color: #000000; background-color:#A4D3EE; border-color: #FFD39B;Width:60%">

Example: 880 = 500 + 200 + 100 + 50 + 20 + 10
<br>
<br>
Input: 880
<br>
<br>
Output: 6
</div>

In [3]:
def no_notes(a):
    Q = [500, 200, 100, 50, 20, 10]
    x = 0
    for i in range(6):
        q = Q[i]
        x += int(a / q)
        a = int(a % q)
    if a > 0:
        x = -1
    return x
print(no_notes(880))
print(no_notes(1000))

6
2


**Câu 104:** Cho mảng A gồm n phần tử. Hãy viết chương trình tạo ra mảng B[1...n], trong đó B[i] là tổng của i phần tử đầu tiên của A.

In [4]:
import random as r
n=int(input("Nhập số lượng phần tử: "))
a=[0]*n
for i in range(n):
    a[i]=r.randint(-300,300)
print("Mảng A",a)
b=[]
for i in range(1,n+1):
    b.append(sum(a[:i]))
print("Mảng B",b)

Nhập số lượng phần tử: 5
Mảng A [180, 116, -254, -206, -190]
Mảng B [180, 296, 42, -164, -354]


**Câu 105:** Cho mảng 2 chiều nxn với các phần tử là các số nguyên. Tìm trong mỗi hàng phần tử lớn nhất rồi đổi chổ nó với phần tử có chỉ số hàng bằng chỉ số cột.

In [5]:
import random as r
n=int(input("Nhập n:"))
a=[]
for i in range(n):
    a.append([])
    for j in range(n):
        t=r.randint(-100,100)
        a[i].append(t)
for i in range(n):
    mx=max(a[i])
    for j in range(n):
        if a[i][j]==mx:
            a[i][j]=a[i][i]
            a[i][i]=mx
print("Mảng mới là:")
for i in range(n):
    for j in range(n):
        print(a[i][j],end=' ')
    print()

Nhập n:4
Mảng mới là:
67 -46 -29 26 
-14 68 14 -91 
-4 -79 98 -73 
10 -13 -61 11 


**Câu 106:** Viết hàm tính toán tổng hiệu các cặp của từng phần tử trong mảng.

<div style="padding: 15px; border: 2px solid transparent; border-color: transparent; margin-bottom: 5px; margin-top: 10px;border-radius: 2px; color: #000000; background-color:#A4D3EE; border-color: #FFD39B;Width:60%">

Example: Mảng [1, 2, 3]
<br>
<br>
Hiệu 1 - 3 = -2 lấy tuyệt đối = 2
<br>
<br>
Hiệu 1 - 2 = -1 lấy tuyệt đối = 1
<br>
<br>
Hiệu 2 - 3 = -1 lấy tuyệt đối = 1
<br>
<br>
Tổng 2 + 1 + 1 = 4
<br>
<br>
Input: [1, 2, 3]
<br>
<br>
Output: 4
</div>

In [6]:
def sum_distinct_pairs(arr):
    result = 0
    i = 0
    while i<len(arr):
        result+=i*arr[i]-(len(arr)-i-1)*arr[i]
        i+=1
    return result
sum_distinct_pairs([1,2,3])

4

**Câu 107:** Viết hàm kiểm tra mảng một chiều có chứa giá trị k không? Nếu có trả về True, ngược lại trả về False.

In [7]:
def FindKArray(A,k):
    for i in range(len(A)):
        if A[i]==k:
            return True
            break
    return False
FindKArray([3,4,5],4)

True

 **Câu 108:** Cho năm số nguyên dương, hãy tìm các giá trị nhỏ nhất và lớn nhất có thể được tính bằng cách tính tổng chính xác bốn trong năm số nguyên. Sau đó, in các giá trị tối thiểu và tối đa tương ứng dưới dạng một dòng gồm hai số nguyên dài được phân tách bằng dấu cách.

In [8]:
def miniMaxSum(arr):
    Max=max(arr)
    Min=min(arr)
    SumMin,SumMax=0,0
    if Max==Min:
        SumMax=SumMin=(len(arr)-1)*Max
    for i in arr:
        if i < Max:
            SumMin+=i
    for i in arr:
        if i>Min:
            SumMax+=i
    print(SumMin,SumMax,sep=' ')
arr=[]
print("Nhập vào 5 số cách nhau bởi dấu cách")
for i in input().rstrip().split():
    arr.append(int(i))
miniMaxSum(arr)

Nhập vào 5 số cách nhau bởi dấu cách
12 14 4 5 9
30 40


**Câu 109:**  Chúng ta có thể xác định một cách đệ quy số lượng tổ hợp của m trong số n,
ký hiệu C (n, m), cho n> = 1 và 0 <= m <= n

Với:

C (n, m) = 1 nếu m = 0 hoặc m = n.

C (n, m) = C (n-1, m) + C (n-1, m-1) nếu 0 <m <n.

Viết hàm đệ quy recursC để tính C(n,m).

In [9]:
def recursC(n,m):
    if m==1 or m==n :
        return 1
    else:
        return recursC(n-1,m)+ recursC(n-1,m-1)
recursC(5,2)    

4

**Câu 110:** Viết chương trình đoán số tiếp theo của một dãy số cấp số cộng hoặc cấp số nhân.
<div style="padding: 15px; border: 2px solid transparent; border-color: transparent; margin-bottom: 5px; margin-top: 10px;border-radius: 2px; color: #000000; background-color:#A4D3EE; border-color: #FFD39B;Width:60%">

Example:
<br>
<br>
<strong>Case 1</strong>
<br>
<br>
Input: [2, 4, 6]
<br>
<br>
Output: 8
<br>
<br>
<strong>Case 2</strong>
<br>
<br>
Input: [2, 6, 18]
<br>
<br>
Output: 54
</div>

In [10]:
def ap_gp_sequence(arr):
    if arr[0]==arr[1]==arr[2]==0:
        return "Số bạn nhập sai"
    else:
        if arr[1]-arr[0]==arr[2]-arr[1]:
            n=2*arr[2]-arr[1]
            return "Cấp số cộng, "+'số tiếp theo của dãy là: '+str(n)
        else:
            n=arr[2]**2/arr[1]
            return "Cấp số nhân, " + 'số tiếp theo của dãy là:  '+str(n)

print(ap_gp_sequence([1,2,3]))
print(ap_gp_sequence([2,6,18]))
print(ap_gp_sequence([0,0,0]))

Cấp số cộng, số tiếp theo của dãy là: 4
Cấp số nhân, số tiếp theo của dãy là:  54.0
Số bạn nhập sai


**Câu 111:** Viết hàm nhập bàn phím n là số nguyên, tạo chuỗi từ 0 đến n hiển thị ra màn hình cách nhau bởi dấu cách.

In [11]:
n=int(input("Nhập n:"))
def printstring(n):
    print("Chuỗi từ 0 đến {}:".format(n))
    for i in range(n+1):
        print(str(i),end=' ')
printstring(n)

Nhập n:4
Chuỗi từ 0 đến 4:
0 1 2 3 4 

**Câu 112:**  Hãy định nghĩa các biến A = 70.0, B = 3 sau và lần lượt in kết quả các toán tử sau:
    
- Example += operator A = ???
- Example -= operator B = ???
- Example *= operator A = ???
- Example /= operator B = ???
- Example %= operator A = ???

In [12]:
a=70;b=3
print("+= operator A =",a+b)
print("-= operator A =",a-b)
print("*= operator A =",a*b)
print("/= operator A =",a//b)
print("%= operator A =",a%b)

+= operator A = 73
-= operator A = 67
*= operator A = 210
/= operator A = 23
%= operator A = 1


**Câu 113:** Viết hàm kiểm tra mảng một chiều có chứa số chẵn không? Nếu có trả về True, ngược lại trả về False.

In [13]:
def EvenNumberIncludingArray(A):
    for i in A:
        if i%2==0:
            return True
        else:
            return False
EvenNumberIncludingArray([6,6,8])

True

**Câu 114:**  Định nghĩa một hàm có input là 2 chuỗi và in chuỗi có độ dài lớn hơn trong giao diện điều khiển. Nếu 2 chuỗi có chiều dài như nhau thì in tất cả các chuỗi theo dòng.

In [14]:
def printstringmax(s1,s2): 
    len1 = len(s1) 
    len2 = len(s2)
    if len1>len2:
        print (s1)
    elif len2>len1:
        print (s2)
    else:
        print(s1)
        print (s2) 
printstringmax("xoai","cam")

xoai


**Câu 115:** Viết hàm kiểm tra danh sách các chuỗi có phải là đối xứng hay không.
<div style="padding: 15px; border: 2px solid transparent; border-color: transparent; margin-bottom: 5px; margin-top: 10px;border-radius: 2px; color: #000000; background-color:#A4D3EE; border-color: #FFD39B;Width:60%">

Example:
<br>
<br>
Input: ['palindrome', 'madamimadam', '', 'foo', 'eyes']
<br>
<br>
Output: [False, True, True, False, False]

</div>

In [15]:
def test_palind(seq):
    results=[]
    for s in seq:
        if s==s[::-1]:
            results.append(True)
        else:
            results.append(False)
    return results
strs = ['palindrome', 'madamimadam', '', 'foo', 'eyes']
test_palind(strs)

[False, True, True, False, False]

**Câu 116:** Viết hàm tìm index các số mà trong danh sách nhỏ hơn 1 ngưỡng threshold <100.
<div style="padding: 15px; border: 2px solid transparent; border-color: transparent; margin-bottom: 5px; margin-top: 10px;border-radius: 2px; color: #000000; background-color:#A4D3EE; border-color: #FFD39B;Width:60%">

Example:
<br>
<br>
Input: [0, 12, 45, 3, 4923, 322, 105, 29, 15, 39, 55]
<br>
<br>
Output: [0, 1, 2, 3, 7, 8, 9, 10]

</div>

In [16]:
def find_index_thres(array,threshold):
    results=[]
    for i in range(len(array)):
        if array[i]<threshold:
            results.append(i)
    return results
arr=[0, 12, 45, 3, 4923, 322, 105, 29, 15, 39, 55]
thres = 100
find_index_thres(arr,thres)

[0, 1, 2, 3, 7, 8, 9, 10]

**Câu 117:** Sử dụng một danh sách để lọc các số lẻ từ danh sách được người dùng nhập vào cách nhau bởi dấu phảy.

<div style="padding: 15px; border: 2px solid transparent; border-color: transparent; margin-bottom: 5px; margin-top: 10px;border-radius: 2px; color: #000000; background-color:#A4D3EE; border-color: #FFD39B;Width:60%">

Example:
<br>
<br>
Input: 1,2,3,4,5,6,7,8,9,10
<br>
<br>
Output: 1,3,5,7,9

</div>

In [17]:
numbers=input("Nhập dãy số của bạn và cách nhau bởi dấu phẩy:")
for num in numbers.split(','):
    if int(num)%2!=0:
        print(num,end=',')

Nhập dãy số của bạn và cách nhau bởi dấu phẩy:1,2,3,4,5,6,7,8,9,10
1,3,5,7,9,

**Câu 118:** Vui lòng viết hàm để xuất một số ngẫu nhiên, chia hết cho 5 và 7, từ 0 đến 200. 

In [19]:
import random
def sample(n):
    results=[]
    for i in range(n):
        if i%5==0 and i%7==0:
             results.append(i)
    return results
random.choice(sample(200))

0

**Câu 119:** Viết chương trình tạo ngẫu nhiên list gồm 5 số chẵn nằm trong đoạn [100;200].

In [20]:
import random
results=[]
for i in range(100,201):
    if i%2==0:
        results.append(i)
random.sample(results,5)

[132, 150, 148, 128, 100]

**Câu 120:** Viết một chương trình tính $\frac{1}{2} + \frac{2}{3} + \frac{3}{4} + ... + \frac{n}{n + 1}$ với một n là số được nhập vào (n> 0).

<div style="padding: 15px; border: 2px solid transparent; border-color: transparent; margin-bottom: 5px; margin-top: 10px;border-radius: 2px; color: #000000; background-color:#A4D3EE; border-color: #FFD39B;Width:60%">

Example:
<br>
<br>
Input: 3
<br>
<br>
Output: 1.92

</div>

In [21]:
n=int(input("Nhập số n >0: "))
Sum=0
for i in range(1,n+1):
    Sum += float(float(i)/(i+1))
round(Sum,2)

Nhập số n >0: 3


1.92

**Câu 121:** In chuỗi Unicode "Hello world".

In [22]:
unicodeString = u"Hello world!"
print (unicodeString)

Hello world!


**Câu 122:** Viết chương trình kiểm tra 3 số có tạo thành tam giác vuông hay không (với x, y, z được nhập từ bàn phím cách nhau bởi dấu cách).

In [24]:
values=input().split()
list_num=[]
for l in values:
    list_num.append(float(l))
x,y,z=sorted(list_num)
if x**2+y**2==z**2:
    print("Ba số {},{},{} tạo thành tam giác vuông".format(x,y,z))
else:
    print("Ba số {},{},{} không tạo thành tam giác vuông".format(x,y,z))

3 4 5
Ba số 3.0,4.0,5.0 tạo thành tam giá vuông


**Câu 123:** Viết chương trình tìm số các số để hình thành phép tính nhân tạo ra n.
<div style="padding: 15px; border: 2px solid transparent; border-color: transparent; margin-bottom: 5px; margin-top: 10px;border-radius: 2px; color: #000000; background-color:#A4D3EE; border-color: #FFD39B;Width:60%">

Example: 12 = 12x1, 12 = 6x2, 12 = 3x4
<br>
<br>
Input: 12
<br>
<br>
Output: 6
</div>

In [25]:
def divisor(n):
    results=[]
    for i in range(1,n+1):
        if not n%i:
            results.append(i)
    return len(results)
divisor(12)

6

**Câu 124:** Viết hàm để lấy các index ở những vị trí có chữ thường của chuỗi.
<div style="padding: 15px; border: 2px solid transparent; border-color: transparent; margin-bottom: 5px; margin-top: 10px;border-radius: 2px; color: #000000; background-color:#A4D3EE; border-color: #FFD39B;Width:60%">

Example:
<br>
<br>
Input: Python
<br>
<br>
Output: [1, 2, 3, 4, 5]

</div>

In [27]:
def test_lower(seq):
    results=[]
    for i in range(len(seq)):
        if seq[i].islower():
            results=results+[i]
        else:
            pass
    return results
test_lower("Python")

[1, 2, 3, 4, 5]

**Câu 125:**  Viết chương trình tính khoảng cách giữa hai số nguyên.
<div style="padding: 15px; border: 2px solid transparent; border-color: transparent; margin-bottom: 5px; margin-top: 10px;border-radius: 2px; color: #000000; background-color:#A4D3EE; border-color: #FFD39B;Width:60%">

Example: 
<br>
<br>
Input: 123 và 256
<br>
<br>
Output: |1-2| + |2-5| + |3-6| = 7
</div>

In [28]:
def digit_distance_nums(a, b):
    dis=abs(a-b)
    dis_str=str(dis)
    results=[]
    for i in dis_str:
        results.append(int(i))
    return sum(results)
digit_distance_nums(123, 256)

7

**Câu 126:** Viết chương trình kiểm tra hai số có là số nguyên tố cùng nhau.

In [29]:
def gcd(p,q):
    while q != 0:
        p, q = q, p%q
    return p
#Ước chung lớn nhất hai số phải bằng 1
def is_coprime(x, y):
    return gcd(x, y) == 1
print(is_coprime(17, 13))
print(is_coprime(17, 21))
print(is_coprime(15, 21))

True
True
False


**Câu 127:** Tính tổng của dãy số âm và dãy số dương trong một mảng rồi so sánh giá trị tuyệt đối nào lớn hơn.
<div style="padding: 15px; border: 2px solid transparent; border-color: transparent; margin-bottom: 5px; margin-top: 10px;border-radius: 2px; color: #000000; background-color:#A4D3EE; border-color: #FFD39B;Width:60%">

Example: Sum positive = 105, Sum negative = -60 
<br>
<br>  
Input: { 0, -10, -11, -12, -13, -14, 15, 16, 17, 18, 19, 20 } 
<br>
<br>
Output: 105


</div>

In [30]:
def test(lst):
    pos_sum = 0
    neg_sum = 0
    for n in lst:
        if n > 0:
            pos_sum += n
        elif n < 0:
            neg_sum += n
    return max(pos_sum, neg_sum, key=abs)

nums = { 0, -10, -11, -12, -13, -14, 15, 16, 17, 18, 19, 20 };
print("Mảng gốc:");
print(nums)
print("Tổng lớn nhất:", test(nums));
nums = { -11, -22, -44, 0, 3, 4 , 5, 9 };
print("Mảng gốc:");
print(nums)
print("Tổng lớn nhất:", test(nums));

Mảng gốc:
{0, 15, 16, 17, -14, -13, -12, -11, -10, 18, 19, 20}
Tổng lớn nhất: 105
Mảng gốc:
{0, 3, 4, 5, 9, -22, -44, -11}
Tổng lớn nhất: -77


**Câu 128:** Viết chương trình xác định vùng của kiểu dữ liệu int, float.

In [31]:
import sys
print("Float value information: ",sys.float_info)
print("\nInteger value information: ",sys.int_info)
print("\nMaximum size of an integer: ",sys.maxsize) 

Float value information:  sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

Integer value information:  sys.int_info(bits_per_digit=15, sizeof_digit=2)

Maximum size of an integer:  2147483647


**Câu 129:** Với 2 list cho trước: [3,4,2,8,4] và [4,3,0,3,9] hãy viết chương trình để tạo list có phần tử là giao của 2 list đã cho.

In [32]:
listA=set([3,4,2,8,4])
listB=set([4,3,0,3,9])

listA &= listB
li=list(listA)
print (li)

[3, 4]


**Câu 135:** Dãy Fibonacci được tính dựa trên công thức sau:

+ f(n) = 0 nếu n = 0

+ f(n) = 1 nếu n = 1

+ f(n) = f(n-1) + f(n-2) nếu n>1

Hãy viết chương trình sử dụng list comprehension để in dãy Fibonacci dưới dạng tách biệt bằng dấu "," và n được người dùng nhập vào.
<div style="padding: 15px; border: 2px solid transparent; border-color: transparent; margin-bottom: 5px; margin-top: 10px;border-radius: 2px; color: #000000; background-color:#A4D3EE; border-color: #FFD39B;Width:60%">

Example: 
<br>
<br>  
Input: 4
<br>
<br>
Output: 0,1,1,2,3


</div>

In [33]:
def fibo(n):
    if n == 0: 
        return 0
    elif n == 1: 
        return 1
    else: 
        return fibo(n-1)+fibo(n-2)
n=int(input("Nhập số n: "))
results=[]
for x in range(0,n+1):
    results.append(str(fibo(x)))
",".join(results)

Nhập số n: 4


'0,1,1,2,3'

**Câu 151:** Viết chương trình đếm và in số ký tự của chuỗi do người dùng nhập vào.
<div style="padding: 15px; border: 2px solid transparent; border-color: transparent; margin-bottom: 5px; margin-top: 10px;border-radius: 2px; color: #000000; background-color:#A4D3EE; border-color: #FFD39B;Width:60%">

Example:
<br>
<br>
Input: AMDK
<br>
<br>
Output:
<br>
A,1
<br>
M,1
<br>
D,1
<br>
K,1
</div>

In [34]:
dict_results={}
seq=input("Nhập chuỗi cần đếm ký tự")
for s in seq:
    dict_results[s]=dict_results.get(s,0)+1
for k,v in dict_results.items():
    print(k,v,sep=",")

Nhập chuỗi cần đếm ký tựAMDK
A,1
M,1
D,1
K,1


**Câu 132:**  Bạn hãy viết một chương trình để in thời gian thực thi (running time of execution) phép tính "1+1" 100 lần.

In [ ]:
from timeit import Timer
t = Timer("for i in range(100):1+1")
print (t.timeit())

**Câu 133:** Viết chương trình để in một số nguyên ngẫu nhiên từ 1 đến 100.

In [33]:
import random
print (random.randrange(1,101))

31


**Câu 134:**  Viết các lệnh assert để xác minh rằng tất cả các số trong list [2,4,6,8] là chẵn.

In [8]:
li = [2,4,6,8]
for i in li:
    assert i%2==0

**Câu 135:** Viết một chương trình chấp nhận đầu vào là chuỗi các số nhị phân 4 chữ số, phân tách bởi dấu phẩy, kiểm tra xem chúng có chia hết cho 5 không. Sau đó in các số chia hết cho 5 thành dãy phân tách bởi dấu phẩy.

<div style="padding: 15px; border: 2px solid transparent; border-color: transparent; margin-bottom: 5px; margin-top: 10px;border-radius: 2px; color: #000000; background-color:#A4D3EE; border-color: #FFD39B;Width:60%">

Example:
<br>
<br>
Input: 0100,0011,1010,1001
<br>
<br>
Output: 1010

</div>

In [7]:
values=input("Nhập các số nhị phân cách nhau bởi dấu phẩy:").split(",")
results=[]
for v in values:
    int_values=int(v,2)
    if int_values%5==0:
        results.append(str(v))
','.join(results)

Nhập các số nhị phân cách nhau bởi dấu phẩy:0100,0011,1010,1001


'1010'

**Câu 136:** Viết hàm tìm ra số unique trong một mảng số nguyên.

In [9]:
def lonelyinteger(arr):
    result=[]
    ans=0
    for a in arr :
        count=arr.count(a)
        result.append(count)
    ans=result.index(1)
    return arr[ans]
lonelyinteger([1,2,3,4,3,2,1,3])

4

**Câu 137:** Viết chương trình hỗ trợ tính cước taxi.

Biết rằng phí dịch vu taxi được tính theo những trường hợp dưới đây :

– Với chiều đường đi dưới 10km, giá phải trả là 10 000 VND/km,

– Từ cây số thứ 11 đến cây số thứ 30, giá phải trả là 6 000 VND/km,

– Kể từ cây số thứ 31, giá phải trả là 4 000 VND/km.
 

In [37]:
def Charge(d):
    if d<=10:
        p=10000*d
        print("Cước phí là",p)
    if 11<=d<=30:
        p=10000*10+6000*(d-10)
        print("Cước phí là",p)
    if d>=31:
        p=10000*10+6000*20+4000*(d-30)
        print("Cước phí là",p)
Charge(23)

Cước phí là 178000


**Câu 138:** Kiểm tra một ma trận có phải là ma trận thưa hay không, chúng ta chỉ cần kiểm tra tổng số phần tử bằng không.
Ma trận là ma trận thưa nếu T ≥ ((m * n) / 2); trong đó T xác định tổng số phần tử bằng không. Nếu là ma trận thưa thì xuất ra chuỗi "Is a Sparse matrix.", ngược lại "Is not Sparse matrix."

In [38]:
def sparsematrix(A):
    count_zero=0
    for i in range(len(A)):
        for j in range(len(A[i])):
            if A[i][j]==0:
                count_zero+=1
    if count_zero>=(len(A)*len(A[i]))//2:
        print("Is a Sparse matrix.")
    else:
        print("Is not Sparse matrix.")
sparsematrix([[0,3,0,0],[2,0,0,0]])

Is a Sparse matrix.


**Câu 139:** Một ma trận A được gọi là tam giác dưới (lower triangular) nếu tất cả các phần tử bên trên đường chéo chính là bằng 0. Hiện thực hàm lowerTriangular trả về true nếu ma trận là tam giác dưới, và false cho ngược lại.

In [39]:
def lower_triangular(A):
    count=0
    for i in range(len(A)):
        count+=1
        for j in range(count,len(A[i])):
            if A[i][j]==0:
                return True
    return False
lower_triangular([[1,0,0],[2,2,0],[3,3,3]])                

True

**Câu 140:** Xác định phần tử đầu tiên có giá trị là số nguyên tố và nhỏ hơn 100; và trả về vị trí của phần tử đó, ngược lại trả về -1.

In [40]:
def prime(n):
    if n<2:
        return True
    else:
        for i in range(2,n):
            if n%i==0:
                return False
    return True
def PrimeNumberIncludingArray(A):
    for i in range(len(A)):
        if prime(A[i]):
            return True
    return False
PrimeNumberIncludingArray([1,3,4,5])

True

**Câu 141:** Hãy hoàn thành định nghĩa hàm deleteAllCharacter dùng để xóa hết các kí tự key xuất hiện trong chuỗi str.

<div style="padding: 15px; border: 2px solid transparent; border-color: transparent; margin-bottom: 5px; margin-top: 10px;border-radius: 2px; color: #000000; background-color:#A4D3EE; border-color: #FFD39B;Width:60%">

Example:
<br>
<br>
Input: string = signal, key = i
<br>
<br>
Output: ignal

</div>

In [43]:
def deleteCharacter(string,key):
    for i in range(len(string)):
        if string[i]==key:
            for j in range(len(string)-1):
                print(string[j+1],end='')
                i-=1
deleteCharacter('signal','i')

ignal

**Câu 142:** Hiện thực hàm tính tổng các dòng lẻ của matrix.

In [44]:
def SumOddRowsMatrix(A):
    Sum=0
    for i in range(1,len(A),2):
        for j in range(len(A[i])):
            Sum+=A[i][j]
    return Sum
SumOddRowsMatrix([[3,4,5],[2,4,3]])

9

**Câu 143:**  Viết một hàm tính tổng các phần tử nằm phía trên đường chéo chính (bao gồm đường chéo chính) ma trận vuông kích thước nxn kiểu số thực và trả về giá trị đó.
<div style="padding: 15px; border: 2px solid transparent; border-color: transparent; margin-bottom: 5px; margin-top: 10px;border-radius: 2px; color: #000000; background-color:#A4D3EE; border-color: #FFD39B;Width:60%">

Example:
<br>
<br>
Input: [[1,3,4],[4,5,6],[3,4,3]]
<br>
<br>
Output: 9

</div>

In [45]:
def trace_matrix(A):
    trace=0
    for i in range(len(A)):
        trace+=A[i][i]
    print(trace)
trace_matrix([[1,3,4],[4,5,6],[3,4,3]])

9


**Câu 144:** Xác định số lượng phần tử có giá trị lẻ trong mảng 1 chiều.

In [46]:
def Odd_NumArray(A):
    d=0
    for i in range(len(A)):
        if A[i]%2!=0:
            d+=1
    print(d)
Odd_NumArray([3,4,5,7,8])

3


**Câu 145:** Đếm xem có bao nhiêu phẩn tử chẳn trong matrix 2 chiều.

In [47]:
def even_numMatrix(A):
    count=0
    for i in range(len(A)):
        for j in range(len(A[i])):
            if A[i][j]%2==0:
                count+=1
    print(count)
even_numMatrix([[3,4,5],[2,4]])

3


**Câu 146:** Viết một hàm xuất các phần tử của mảng hai chiều (nxm) ra màn hình không có khoảng trắng hay dấu xuống dòng ở cuối, giữa các phần tử chỉ có đúng 1 khoảng trắng.

In [48]:
def printMatrix(A):
    for i in range(len(A)):
        for j in range(len(A[i])):
            print(A[i][j],end=' ')
        print()
printMatrix([[1,2,3],[3,4,4]])

1 2 3 
3 4 4 


**Câu 147:** Một ma trận vuông A được gọi là đối xứng nếu ma trận A' = A, với A' là ma trận chuyển vị.

Hiện thực hàm isSymmetricMatrix để kiểm tra xem một ma trận vuông có phải là ma trận đối xứng không?

Nếu đối xứng thì trả về 1, ngược lại trả về 0.

In [49]:
def isSymmetricMatrix(A):
    Sum=0
    for i in range(len(A)):
        for j in range(len(A[i])):
            if A[i][j]==A[j][i]:
                Sum+=1
    if Sum==len(A)*len(A[i]):
        return 1
    else:
        return 0
isSymmetricMatrix([[1,7,3],[7,4,5],[3,5,6]])

1

**Câu 148:** Viết chương trình hiển thị True nếu ba đối số theo thứ tự không giảm;
mặt khác, nó trả về False.

Ví dụ: thứ tự (1,2, 3) và in_order (1, 2, 2) đều trả về true, trong khi thứ tự (1, 3,2) trả về false.

In [50]:
def inorder(a,b,c):
    if a<b and b<=c:
        return True
    else:
        return False
inorder(1,3,2)

False

**Câu 148:** Xác định phần tử chẵn thứ k trong mảng và trả về giá trị của phần tử đó (trả về “-1” nếu không tồn tại).
<div style="padding: 15px; border: 2px solid transparent; border-color: transparent; margin-bottom: 5px; margin-top: 10px;border-radius: 2px; color: #000000; background-color:#A4D3EE; border-color: #FFD39B;Width:60%">

Example:
<br>
<br>
Input: [3,4,5,2]
<br>
<br>
Output: [-1, 4, -1, 2]

</div>

In [51]:
B=list()
def Even_Array(A):
    for i in range(len(A)):
        if A[i]%2==0:
            B.append(A[i])
        else:
            B.append(-1)
    print(B)
Even_Array([3,4,5,2])

[-1, 4, -1, 2]


**Câu 149:** Viết một hàm tính tổng các phần tử trong ma trận (nxm) kiểu số thực và trả về giá trị đó.

In [52]:
def sumMatrix(A):
    Sum=0
    for i in range(len(A)):
        for j in range(len(A[i])):
            Sum+=A[i][j]
    print(Sum)
sumMatrix([[3,4,3],[9,4,5]])

28


**Câu 150:** Hãy tính tổng của hai ma trận 2 chiều có cùng kích thước.

In [53]:
def add_Matrix(A,B):
    for i in range(len(A)):
        for j in range(len(A[i])):
            A[i][j]=A[i][j]+B[i][j]
    print(A)
add_Matrix([[1,2,3],[2,3,4]],[[4,3,2],[3,2,3]])

[[5, 5, 5], [5, 5, 7]]


----